pkg/uprobetracer/ldcache_parser.go

```diff
@@ -0,0 +1,210 @@
+		return errors.New("reading from bytes: length mismatched")
+	}
+	buffer := bytes.NewBuffer(rawData)
+	err := binary.Read(buffer, binary.NativeEndian, obj)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func readStringFromBytes(data []byte, startPos uint32) string {
+	res := ""
+	for i := startPos; i < uint32(len(data)); i++ {
+		if data[i] == 0 {
+			return res
+		}
+		res += string(data[i])
+	}
+	return ""
+}
+
+func readCacheFormat1(data []byte) []ldEntry {
+	var ldEntries []ldEntry
+
+	ldCache := ldCache1{}
+	if uint32(len(data)) <= ldCache1Size {
+		return nil
+	}
+	err := readFromBytes(&ldCache, data[:ldCache1Size])
+	if err != nil {
+		return nil
+	}
+	ldEntriesOffset := ldCache1Size
+	ldStringsOffset := ldCache1Size + ldCache1EntrySize*ldCache.EntryCount
+	for i := uint32(0); i < ldCache.EntryCount; i++ {
+		entryOffset := ldEntriesOffset + i*ldCache1EntrySize
+		entry := ldCache1Entry{}
+		if uint32(len(data)) <= entryOffset+ldCache1EntrySize {
+			return nil
+		}
+		err := readFromBytes(&entry, data[entryOffset:entryOffset+ldCache1EntrySize])
+		if err != nil {
+			return nil
+		}
+		keyOffset := ldStringsOffset + entry.Key
+		valueOffset := ldStringsOffset + entry.Value
+		key := readStringFromBytes(data, keyOffset)
+		value := readStringFromBytes(data, valueOffset)
+		ldEntries = append(ldEntries, ldEntry{key, value})
+	}
+	return ldEntries
+}
+
+func readCacheFormat2(data []byte) []ldEntry {
+	var ldEntries []ldEntry
+
+	if !bytes.Equal([]byte(cache2Header), data[:len(cache2Header)]) {
+		return nil
+	}
+	ldCache := ldCache2{}
+	if uint32(len(data)) <= ldCache2Size {
+		return nil
+	}
+	err := readFromBytes(&ldCache, data[:ldCache2Size])
+	if err != nil {
+		return nil
+	}
+	ldEntriesOffset := ldCache2Size
+	for i := uint32(0); i < ldCache.EntryCount; i++ {
+		entryOffset := ldEntriesOffset + i*ldCache2EntrySize
+		entry := ldCache2Entry{}
+		if uint32(len(data)) <= entryOffset+ldCache2EntrySize {
+			return nil
+		}
+		err := readFromBytes(&entry, data[entryOffset:entryOffset+ldCache2EntrySize])
+		if err != nil {
+			return nil
+		}
+		keyOffset := entry.Key
+		valueOffset := entry.Value
+		key := readStringFromBytes(data, keyOffset)
+		value := readStringFromBytes(data, valueOffset)
+		ldEntries = append(ldEntries, ldEntry{key, value})
+	}
+	return ldEntries
+}
+
+// simulate the loader's behaviour, find library path in containers' `/etc/ld.so.cache`.
+// see https://github.com/bminor/glibc/blob/master/elf/cache.c#L292, the `print_cache` func
+// see https://github.com/iovisor/bcc/blob/master/src/cc/bcc_proc.c#L508, the `bcc_procutils_which_so` func
+func parseLdCache(ldCachePath string, libraryName string) ([]string, error) {
+	fileInfo, err := os.Stat(ldCachePath)
+	if err != nil {
+		return nil, fmt.Errorf("stat file %q: %q", ldCachePath, err.Error())
+	}
+	if fileInfo.Mode()&os.ModeNamedPipe != 0 {
+		return nil, fmt.Errorf("ldCache file %q is a pipe", ldCachePath)
+	}
+	ldCacheFile, err := os.ReadFile(ldCachePath)
```

Comment:
I realize this is still vulnerable to TOCTOU issues: between the `SecureJoin()` and the `os.Stat()`/`os.ReadFile()`, a malicious workload in a container could replace the path by a symlink to `/dev/zero` (leading to `ig` being oomkilled) or to `../../../../etc/shadow` or similar.

I am not sure this can be fixed without `openat2()` in Linux 5.6 and `resolve=RESOLVE_IN_ROOT`.

@rata Wdyt?

Maybe it is good enough for this PR and we address this in a follow-up issue.